### ISSUE TITLE
Bug: object-shorthand does not report "key: () => {}" with "always"

### ISSUE THREAD

**Issue by ilyub** [eslint/eslint#15650]
### Environment

Node version: 14.18.0
npm version: 8.0.0
Local ESLint version: 8.9.0
Global ESLint version: 8.9.0
Operating System: Windows 10

### What parser are you using?

vue-eslint-parser

### What did you do?

```js
/*eslint object-shorthand: ["error", "always"]*/

const foo = {
  f1: () => {},
  f2() {}
};
```

### What did you expect to happen?

One error

### What actually happened?

No error

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ilyub** [comment]
> it would be pretty dangerous to imply that they're interchangeable.

Yet, this rule implies that arrow and shorthand functions are interchangeable in this situation:
```ts
/*eslint object-shorthand: ["error", "always", { avoidExplicitReturnArrows: true }]*/

const foo = {
  "f-1": () => {},
  "f-2"() {}
};
```

![image](https://user-images.githubusercontent.com/70687432/155860826-8ebf3bcd-311a-44f5-bd6b-6aa0c107f99c.png)

Also, if arrow and shorthand are not interchangeable when why this whole rule exists?

**Comment by mdjermanovic** [comment]
@ilyub

> ```js
> /*eslint object-shorthand: ["error", "always"]*/
> 
> const foo = {
>   f1: () => {},
>   f2() {}
> };
> ```

This works as intended, arrow functions are reported only with `avoidExplicitReturnArrows: true `.

Without that option, `"always"` reports only properties that can be shorthand and function expressions.

```js
/*eslint object-shorthand: ["error", "always"]*/

const foo = {
  x: x,
  y: function () {}
};
```

**Comment by ilyub** [comment]
> It seems like the avoidExplicitReturnArrows option is exactly what you need, though.

@ljharb, why do you write this when "avoidExplicitReturnArrows" clearly does not work for my case?
"avoidExplicit<b>Return</b>Arrows" is intended for functions with "return" statement and my sample does not contain returns:
```js
const foo = {
  f1: () => {},
  f2() {}
};
```

**Comment by ljharb** [comment]
@ilyub "explicit return arrows" means "any arrow function where the RHS is a block, and not an expression". the presence of a return statement is irrelevant.

**Comment by ilyub** [comment]
@ljharb, I examined once again and agree that my example above is not right for avoidExplicitReturnArrows.
But here is the right example:
```
/*eslint object-shorthand: ["warn", "always", { avoidExplicitReturnArrows: true }]*/
const foo = {
  fn1(value) {
    return value;
  },
  fn2: value => value
};
```
My point is still the same: This rule does not impose single style in some cases.

Sorry, for bothering you twice after ticket is closed.
If my point does not count, I give up.

**Comment by mdjermanovic** [comment]
> > Without that option, "always" reports only properties that can be shorthand and function expressions.
> 
> I see but I don't understand why it is intended this way. Why the rule intended to impose consistent function style allows two styles in some circumstances?

I think it wasn't clear whether or not this rule should apply to arrow functions at all. The initial version did apply, then that part was reverted before the release (https://github.com/eslint/eslint/issues/2414), and at the end the option was added (https://github.com/eslint/eslint/issues/7564).

Either way, the rule is configurable, so it can be configured to disallow them.

Except for the implicit return form like `foo => bar`. The reasoning in https://github.com/eslint/eslint/issues/7564 was that they're already more concise than methods.

**Comment by ljharb** [comment]
The rule isn’t about conciseness tho, since it can be configured to enforce a longer form.

**Comment by mdjermanovic** [comment]
We're not adding any more options to stylistic rules.

A workaround without autofix is no-restricted-syntax with `"Property > ArrowFunctionExpression.value"` ([Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1zeW50YXg6IFtcImVycm9yXCIsIFwiUHJvcGVydHkgPiBBcnJvd0Z1bmN0aW9uRXhwcmVzc2lvbi52YWx1ZVwiXSAqL1xuXG5jb25zdCBmb28gPSB7XG4gIGZuMSh2YWx1ZSkge1xuICAgIHJldHVybiB2YWx1ZTtcbiAgfSxcbiAgZm4yOiB2YWx1ZSA9PiB2YWx1ZVxufTtcblxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==))

**Comment by ljharb** [comment]
I understand that, but it’s still a gap in the purpose of the rule :-) as always, no-restricted-syntax is insufficient because it’s not targeted, it has no autofix, and it’s not practical to alter the configuration when used in a shared config.
